### PR TITLE
Mesh to STL

### LINKED ISSUE #4: STL_Toolkit: Complex geometry
#### Description:

Currently STL Toolkit is implementing a Geometry > ToPolyline > ToSTL workflow to export a STL file. While it works for most regular geometry, it has limitation on complex geometry. 

A possible solution to complex geometry is to add support to Mesh, which should be relatively straightforward based on current workflow.

### LINKED ISSUE #5: STL_Toolkit: Donut shape geometry
#### Description:

One limitation of the current Geometry > ToPolyline > ToSTL workflow is it cannot handle donut shape geometry, meaning a surface with external edges as well as internal edges. This is very common case in building environment, for example, a window on a wall, as the following example showing. This case is showing a co-planar external edges and internal edges, but for curvature surfaces, the edges may not be on the same plane.
image

A test script is included below.
[STL_Toolkit_TestScript.zip](https://github.com/BHoM/STL_Toolkit/files/4919074/STL_Toolkit_TestScript.zip)

### PR DESCRIPTION
**Pull request by BingWangUS**
### Issues addressed by this PR

Closes #4 
Closes #5 

### Test files

[STLToolkit_#4_Mesh.zip](https://github.com/BHoM/STL_Toolkit/files/5911663/STLToolkit_.4_Mesh.zip)

### Changelog

 - Modify ToPolyline method to handle mesh file

### CODE REVIEW

**Comment by BingWangUS** [comment]
@BHoMBot check compliance

**Comment by BingWangUS** [comment]
@BHoMBot check installer

File: STL_Engine/Convert/ToPolyline.cs [hunk 1]
```diff
@@ -81,6 +81,26 @@ public static List<Polyline> ToPolyline(this PolySurface psurface)
...
+            {
+                List<Point> controlPoints = new List<Point>();
+                controlPoints.Add(mesh.Vertices[face.A]);
+                controlPoints.Add(mesh.Vertices[face.B]);
+                controlPoints.Add(mesh.Vertices[face.C]);
+                if (face.D != -1)
+                {
+                    controlPoints.Add(mesh.Vertices[face.D]);
+                }
+                Polyline polyline = new Polyline() { ControlPoints = controlPoints };
```
Comment (FraserGreenroyd) [inline STL_Engine/Convert/ToPolyline.cs:97]: This won't produce a closed polyline, is that what's after? If we want a closed polyline we need to do:
`controlPoints.Add(controlPoints.First());`
just above this line.

**Review by FraserGreenroyd — COMMENTED** [review]
Comment as a query, might not be relevant if closed polylines aren't needed.

File: STL_Engine/Convert/ToPolyline.cs (hunk 1 above)
Comment (vbeardow) [reply STL_Engine/Convert/ToPolyline.cs:97]: Thanks @FraserGreenroyd - tested and this eliminates the warning that @BingWangUS Bing mentions in the PR 👍

**Comment by BingWangUS** [comment]
Thank you both for helping on this. Resolved now.

@vbeardow @FraserGreenroyd there is a fail for the installer check but it seems not caused by this PR. FYI.

**Comment by BingWangUS** [comment]
@BHoMBot check compliance

**Comment by BingWangUS** [comment]
@BHoMBot check installer

**Review by vbeardow — APPROVED** [review]
LGTM

**Review by FraserGreenroyd — APPROVED** [review]

**Pull request merged**